### DISCUSSION TITLE
Add helper method for deleting typed cookies

### DISCUSSION THREAD

**Discussion by mikeybinns** [sergiodxa/remix-utils#213] [Ideas]
Because values passed into `.serialize()` are type-checked, it's really annoying to delete cookies. You must first parse the set cookie, making sure it's valid just to pass it back to delete the cookie, and if it's invalid (say you pass something like `{} as NonNullable<Awaited<ReturnType<typeof myCookie.parse>>>` like I tried) then it prevents you from deleting the cookie.

I suggest adding a helper method called delete that takes no parameters and essentially skips the serialize typechecking steps, sets `null` as the value and sets the expires to `new Date(1970, 0, 1)`.

Example usage:
```ts
return json({ success: true } as const, {
  headers: new Headers({
    "Set-Cookie": await myCookie.delete(),
  }),
});

// instead of

const currentCookie = await myCookie.parse(request.headers.get("Cookie"));
if (currentCookie === null) {
  return json({ success: true } as const);
}
return json({ success: true } as const, {
  headers: new Headers({
    "Set-Cookie": await myCookie.serialize(currentCookie, {
      expires: new Date(1970, 0, 1),
    }),
  }),
});
```

Search terms:
delete cookie

**Comment by sergiodxa** [comment]
I would suggest to add this directly to Remix so there's a `myCookie.delete()` method in Remix cookies, then I can expose them typed cookies here.

**Reply by mikeybinns** [reply]
I vaguely remember asking for this before and being told it was unnecessary because the HTML recommended way to delete cookies is via an expires header, but I can't find that conversation / issue / discussion.

In remix it's less of an issue because you can set any value in serialize and it's fine, meaning you don't have to bypass a typecheck, you could just do this:
```ts
return json({ success: true } as const, {
  headers: new Headers({
    "Set-Cookie": await myCookie.serialize(null, {
      expires: new Date(1970, 0, 1),
    }),
  }),
});
```

The problem where you need to retrieve the cookie only comes from having to pass a valid value to serialize which is only in this package.

If you like, I can submit a PR, but as it's a new feature request, I'm raising it here first 🙂

**Reply by sergiodxa** [reply]
I [updated the docs](https://github.com/sergiodxa/remix-utils/commit/802b6034c2c3aee338d9350209a5316669a0c324) to mention to to delete a typed cookie using `.nullable()` in your schema to let you pass `null` to serialize.

This what I always do on my own code, and even if you don't plan to delete them it's good idea since cookies will return `null` if they are not in the request's Cookies header.

I still think that having a `myCookie.destroy` method on Remix's Cookie object itself is still useful, I recommend you to propose it again and suggest that it could be a helper to set the expires to a past date since lots of people doesn't know they can do that.

**Reply by mikeybinns** [reply]
Cool, that makes a lot of sense.
I've added a feature discussion here for remix to add a delete method to cookies: https://github.com/remix-run/remix/discussions/6934
